onnxruntime/contrib_ops/webgpu/quantization/matmul_nbits.cc

```diff
@@ -682,6 +233,156 @@ fn dequantize_packed8xU4(packed_value : u32, zero_point : output_element_t, scal
   return Status::OK();
 }
 
+// Apply similar idea with DP4AMatMulNBitsSmallMProgram algorithm.
+Status MatMulNBitsProgram::GenerateShaderCode(ShaderHelper& shader) const {
+  const auto& a = shader.AddInput("input_a", ShaderUsage::UseValueTypeAlias);
+  const auto& b = shader.AddInput("input_b");
+  shader.AddInput("scales_b");
+  if (has_zero_points_) {
+    shader.AddInput("zero_points", ShaderUsage::UseUniform);
+  }
+  shader.AddOutput("output", ShaderUsage::UseElementTypeAlias);
+  const uint32_t components_a = a.NumComponents();
+  const uint32_t components_b = b.NumComponents() / 4;  // b is stored as uint32 which includs 4 uint8.
+  constexpr uint32_t tile_size_k_vec = 16;
```

Comment:
does this mean that elements_in_vec has to be 2, because of elements_in_vec is 4. tile_k_size would be 64 - then we would need to have 2 scales,zeropoints per cycle if our block size is 32.